--- esp_http_client.c.orig	2021-09-02 02:15:36.000000000 -0700
+++ esp_http_client.c	2021-11-16 06:19:12.242226552 -0800
@@ -1050,7 +1050,7 @@
                 }
                 http_dispatch_event(client, HTTP_EVENT_ON_FINISH, NULL, 0);
 
-                client->response->buffer->raw_len = 0;
+                //client->response->buffer->raw_len = 0;    // patch_9 A.Chiebao 20211116
                 if (!http_should_keep_alive(client->parser)) {
                     ESP_LOGD(TAG, "Close connection");
                     esp_http_client_close(client);
